src/libsystemd/sd-event/sd-event.c

```diff
@@ -3184,6 +3191,42 @@ static int process_inotify(sd_event *e) {
         return done;
 }
 
+static int source_trigger_ratelimit(sd_event_source *s) {
+        int r;
+
+        assert(s);
+
+        if (!s->ratelimit)
+                return false;
+
+        if (s->ratelimit->effective)
+                return true;
+
+        if (ratelimit_below(&s->ratelimit->ratelimit))
+                return false;
+
+        /* Let's schedule high priority wake up after interval elapsed */
+        r = sd_event_source_set_time(s->ratelimit->timeout_source, s->ratelimit->ratelimit.begin + s->ratelimit->ratelimit.interval);
+        if (r < 0)
+                return r;
+
+        r = sd_event_source_set_priority(s->ratelimit->timeout_source, SD_EVENT_PRIORITY_IMPORTANT);
```

Comment:
i still think this is wrong: this should have the same prio as the original event source, not be higher. If some event source is set to some really high prio work on them should not be deferred just because the user installed a ton of event sources with ratelimits that run out

Reply:
(oh, and if the main event source changes prio — because somebody calls sd_event_source_set_priority() on it — the associated rate limit timer event should change prio too. this is what i mean by "recursive" btw, which i don#t like, i.e. al the props from the main event source always have to be propagated to the rate limit timer event source. I'd much rather have it that the primary event source acts as timer some times and as IO event source in others, depending if ratelimit is on or onot)